File: TLS Inspector/Supporting Files/Frameworks/App Links/AppLinks.m
#import "AppLinks.h"
#import "UIDevice+PlatformString.h"
@import StoreKit;
@import MessageUI;

@interface AppLinks() <SKStoreProductViewControllerDelegate, MFMailComposeViewControllerDelegate> {
    void (^dismissedBlock)(void);
}

@property (strong, nonatomic) UIViewController * viewController;

@end

@implementation AppLinks

#define APP_LAUNCH_KEY @"__APP_LAUNCH_TIMES"
#define APP_LAUNCH_RATE_KEY @"__APP_LAUNCH_RATE"
#define APP_ID 1100539810
#define APP_NAME @"TLS Inspector"
#define APP_EMAIL @"'TLS Inspector Project Manager' <hello@tlsinspector.com>"

- (void) showAppInAppStorInViewController:(UIViewController *)viewController dismissed:(void (^)(void))dismissed {
    SKStoreProductViewController * productViewController = [SKStoreProductViewController new];
    productViewController.delegate = self;
    NSString * appIDString = [NSString stringWithFormat:@"%lu", (unsigned long)APP_ID];
    [productViewController loadProductWithParameters:@{
                                                       SKStoreProductParameterITunesItemIdentifier:
                                                           appIDString}
                                     completionBlock:nil];
    self.viewController = viewController;
    [viewController presentViewController:productViewController animated:YES completion:nil];
    dismissedBlock = dismissed;
}

- (void) appLaunchRate {
    NSUserDefaults * defaults = [NSUserDefaults standardUserDefaults];
    if (![defaults boolForKey:APP_LAUNCH_RATE_KEY]) {
        NSNumber * launchTimesNumber = [defaults objectForKey:APP_LAUNCH_KEY];
        if (!launchTimesNumber) {
            launchTimesNumber = @1;
        }
        NSUInteger launchTimes = [launchTimesNumber unsignedIntegerValue];
        if (launchTimes > 2 && [SKStoreReviewController class]) {
            [SKStoreReviewController requestReview];
            [defaults setBool:YES forKey:APP_LAUNCH_RATE_KEY];
        }
        launchTimes++;
        [defaults setObject:[NSNumber numberWithUnsignedInteger:launchTimes] forKey:APP_LAUNCH_KEY];
    }
}

- (void)productViewControllerDidFinish:(SKStoreProductViewController *)viewController {
    if (viewController) {
        [self.viewController dismissViewControllerAnimated:YES completion:^{
            if (self->dismissedBlock) {
                self->dismissedBlock();
            }
        }];
    } else {
        if (self->dismissedBlock) {
            self->dismissedBlock();
        }
    }
}

- (void) showEmailComposeSheetForAppInViewController:(UIViewController *)viewController withComments:(NSString *)comments includeLogs:(BOOL)includeLogs dismissed:(void (^)(void))dismissed {
    MFMailComposeViewController * mailController = [MFMailComposeViewController new];
    mailController.mailComposeDelegate = self;
    [mailController setSubject:[NSString stringWithFormat:@"%@ Feedback", APP_NAME]];
    [mailController setToRecipients:@[APP_EMAIL]];

    NSDictionary * infoDict = [[NSBundle mainBundle] infoDictionary];
    NSString * bundleName = [[NSBundle mainBundle] bundleIdentifier];
    NSString * bundleVersion = [infoDict objectForKey:@"CFBundleShortVersionString"];
    NSNumber * build = [infoDict objectForKey:@"CFBundleVersion"];
    NSString * bundleBuild = [NSString stringWithFormat:@"%i", [build intValue]];
    NSString * gitRevision = GIT_REVISION;
    NSString * deviceName = [[UIDevice currentDevice] platformString];
    NSOperatingSystemVersion systemVersion = [[NSProcessInfo processInfo] operatingSystemVersion];
    NSString * body = [NSString stringWithFormat:@"Please do not delete the following line:<br/>%@ %@ (%@ - %@) %@ %li.%li.%li",
                       bundleName, bundleVersion, bundleBuild, gitRevision, deviceName, (long)systemVersion.majorVersion, (long)systemVersion.minorVersion, (long)systemVersion.patchVersion];
    [mailController setMessageBody:[NSString stringWithFormat:@"<p>%@<br/><br/></p><hr/><p><small>%@</small></p>", (comments != nil ? comments : @""), body] isHTML:YES];

    if (includeLogs) {
        NSArray * paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory,NSUserDomainMask, YES);
        NSString * documentsDirectory = [paths objectAtIndex:0];
        NSString * cklogPath = [documentsDirectory stringByAppendingPathComponent:@"CertificateKit.log"];
        NSData * logData = [NSData dataWithContentsOfFile:cklogPath];
        if (logData != nil) {
            [mailController addAttachmentData:logData mimeType:@"text/plain" fileName:@"TLS Inspector.log"];
        }
        NSString * exceptionsLogPath = [documentsDirectory stringByAppendingPathComponent:@"exceptions.log"];
        if ([NSFileManager.defaultManager fileExistsAtPath:exceptionsLogPath]) {
            NSData * exceptionData = [NSData dataWithContentsOfFile:exceptionsLogPath];
            if (exceptionData != nil) {
                [mailController addAttachmentData:exceptionData mimeType:@"text/plain" fileName:@"Exceptions.log"];
            }
        }
    }
    
    if (!mailController) {
        if (dismissed) {
            dismissed();
        }
        return;
    }

    dismissedBlock = dismissed;

    [viewController presentViewController:mailController animated:YES completion:nil];
    self.viewController = viewController;
}

- (void)mailComposeController:(MFMailComposeViewController *)controller didFinishWithResult:(MFMailComposeResult)result error:(nullable NSError *)error {
    [controller dismissViewControllerAnimated:YES completion:^{
        if (self->dismissedBlock) {
            self->dismissedBlock();
        }
    }];
}

@end

File: TLS Inspector/Features/AppLinks.swift
import UIKit
import StoreKit
import MessageUI

class AppLinks : NSObject, SKStoreProductViewControllerDelegate, MFMailComposeViewControllerDelegate {
    // Settings key to track the number of times the application has been launched
    private let appLaunchKey = "__APP_LAUNCH_TIMES"
    // Settings key to track if the user has been prompted to rate the app
    private let appLaunchRateKey = "__APP_LAUNCH_RATE"
    // App Store ID number for TLS Inspector. Not secret.
    private let appID = "1100539810"
    private let appName = "TLS Inspector"
    private let appEmail = "'TLS Inspector Project Manager' <hello@tlsinspector.com>"
    // Used to track which app store views come from the app v.s. which come from our website
    // not used to track you or your device.
    private let appCampaignToken = "acid-burn"
    private var dimissedBlock: (() -> Void)?
    public static var current = AppLinks()

    public func showAppStore(_ viewController: UIViewController, dismissed: (() -> Void)?) {
        let productViewController = SKStoreProductViewController()
        productViewController.delegate = self
        let parameters = [
            SKStoreProductParameterITunesItemIdentifier: appID,
            SKStoreProductParameterCampaignToken: appCampaignToken,
        ]
        productViewController.loadProduct(withParameters: parameters, completionBlock: nil)
        viewController.present(productViewController, animated: true, completion: nil)
        self.dimissedBlock = dismissed
    }

    internal func productViewControllerDidFinish(_ viewController: SKStoreProductViewController) {
        viewController.dismiss(animated: true) {
            self.dimissedBlock?()
        }
    }

    @available(iOS 10.3, *)
    public func appLaunchRate() {
        let defaults = UserDefaults.standard
        if !defaults.bool(forKey: appLaunchRateKey) {
            var launchTimes: Int = 1
            if let times = defaults.object(forKey: appLaunchKey) as? NSNumber {
                launchTimes = times.intValue
            }
            if launchTimes > 2 {
                SKStoreReviewController.requestReview()
                defaults.set(true, forKey: appLaunchRateKey)
            }
            launchTimes += 1
            defaults.set(launchTimes, forKey: appLaunchKey)
        }
    }

    public func showEmailCompose(viewController: UIViewController, object: SupportType, includeLogs: Bool, dismissed: (() -> Void)?) {
        if !MFMailComposeViewController.canSendMail() {
            return
        }

        let mailController = MFMailComposeViewController()
        mailController.mailComposeDelegate = self
        mailController.setSubject(appName + " Feedback")
        mailController.setToRecipients([appEmail])
        mailController.setMessageBody(object.body(), isHTML: true)

        if includeLogs {
            self.attachLogFile(actualName: "CertificateKit.log", attachmentName: "TLS Inspector.log", mailController: mailController)
            self.attachLogFile(actualName: "exceptions.log", attachmentName: "Exceptions.log", mailController: mailController)
        }

        self.dimissedBlock = dismissed
        viewController.present(mailController, animated: true, completion: nil)
    }

    func mailComposeController(_ controller: MFMailComposeViewController, didFinishWith result: MFMailComposeResult, error: Error?) {
        controller.dismiss(animated: true) {
            self.dimissedBlock?()
        }
    }

    private func attachLogFile(actualName: String, attachmentName: String, mailController: MFMailComposeViewController) {
        let paths = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)
        if paths.count == 0 {
            LogError("Error getting documents directory")
            return
        }
        let documentsDirectory = URL(fileURLWithPath: paths[0])

        let ckLogPath = documentsDirectory.appendingPathComponent(actualName)
        if !FileManager.default.fileExists(atPath: ckLogPath.path) {
            LogError("Error getting CertificateKit.log")
            return
        }

        var data: Data?
        do {
            try data = Data(contentsOf: ckLogPath)
        } catch {
            LogError("Error reading log file: \(error.localizedDescription)")
            return
        }

        mailController.addAttachmentData(data ?? Data(), mimeType: "text/plain", fileName: attachmentName)
    }
}
